pkg/client/delegation/redelegation.go

```diff
@@ -27,22 +38,54 @@ func (d redelegation) GetGatewayAddress() string {
 }
 
 // newRedelegationEventFactoryFn is a factory function that returns a
-// function that attempts to deserialise the given bytes into a redelegation
+// function that attempts to deserialize the given bytes into a redelegation
 // struct. If the delegate struct has an empty app address then an
 // ErrUnmarshalRedelegation error is returned. Otherwise if deserialisation
 // fails then the error is returned.
 func newRedelegationEventFactoryFn() events.NewEventsFn[client.Redelegation] {
-	return func(redelegationEventBz []byte) (client.Redelegation, error) {
-		redelegationEvent := new(redelegation)
-		if err := json.Unmarshal(redelegationEventBz, redelegationEvent); err != nil {
+	return func(eventBz []byte) (client.Redelegation, error) {
+		txEvent := new(TxEvent)
+		// Try to deserialize the provided bytes into a TxEvent.
+		if err := json.Unmarshal(eventBz, txEvent); err != nil {
 			return nil, err
 		}
-
-		if redelegationEvent.AppAddress == "" || redelegationEvent.GatewayAddress == "" {
-			return nil, events.ErrEventsUnmarshalEvent.
-				Wrapf("with redelegation: %s", string(redelegationEventBz))
+		// Check if the TxEvent has empty transaction bytes, which indicates
+		// the message might not be a valid transaction event.
+		if bytes.Equal(txEvent.Tx, []byte{}) {
+			return nil, events.ErrEventsUnmarshalEvent.Wrapf("%s", string(eventBz))
 		}
-
-		return redelegationEvent, nil
+		// Iterate through the log entries to find EventRedelegation
+		for _, event := range txEvent.Result.Events {
+			if event.GetType_() == "pocket.application.EventRedelegation" {
+				var redelegationEvent redelegation
+				for _, attr := range event.Attributes {
+					switch attr.Key {
+					case "app_address":
+						appAddr, err := unescape(attr.Value)
+						if err != nil {
+							return nil, events.ErrEventsUnmarshalEvent.Wrapf("%s", string(eventBz))
+						}
+						redelegationEvent.AppAddress = appAddr
+					case "gateway_address":
+						gatewayAddr, err := unescape(attr.Value)
+						if err != nil {
+							return nil, events.ErrEventsUnmarshalEvent.Wrapf("%s", string(eventBz))
+						}
+						redelegationEvent.GatewayAddress = gatewayAddr
+					}
+				}
+				// Handle the redelegation event
+				if redelegationEvent.AppAddress == "" || redelegationEvent.GatewayAddress == "" {
+					return nil, events.ErrEventsUnmarshalEvent.
+						Wrapf("%s", string(eventBz))
```

Comment:
Ditto - more details on the error. 

Future devs will likely not put the effort into making errors more meaningful.